src/main/java/edu/wpi/grip/core/Connection.java

```diff
@@ -29,6 +30,10 @@ public Connection(EventBus eventBus, Socket<? extends T> outputSocket, Socket<T>
         checkNotNull(inputSocket);
         checkNotNull(outputSocket);
         checkNotNull(eventBus);
+        checkArgument(!Socket.Direction.INPUT.equals(outputSocket.getDirection()),
+                "outputSocket cannot be an input socket");
+        checkArgument(!Socket.Direction.OUTPUT.equals(inputSocket.getDirection()),
+                "inputSocket cannot be an output socket");
```

Comment:
* Sockets know nothing about connections or the direction they go.  Putting that logic in the socket class would be leaking `Connection`'s implementation details into somewhere they don't belong
* The `Direction` enum and the fact that sockets have a direction is already public.  Adding a getter doesn't leak anything.
* Again, you're not going to convince me that getters are evil, and there's no way we're going through the entire codebase and removing all of the getters.  All of the other fields in `Socket` and `SocketHint` have getters, so this one does too.